Check Jenkins job configs into git

spxtr opened the pull request:
@ixdy @ihmccreery 

There is a docker container running on jenkins-master named job-builder that was built out of `hack/jenkins/job-builder-image`. Its purpose is to run [Jenkins Job Builder](http://docs.openstack.org/infra/jenkins-job-builder/) which can take a directory full of YAML files, convert them to XML, and post only the ones that changed to Jenkins. The YAML files are easy to read/write, and Jenkins Job Builder supports templates, macros, and editable defaults. I put an example config in the PR, that's the one that will update jobs as needed.

The container can live anywhere, it doesn't have to be on Jenkins, but as long as we have multijobs it's much simpler for it to live on Jenkins.

Updating a single running job doesn't affect the current build. We need to be careful about multijobs, since it'd be awkward to update the entire batch of jobs while it's halfway through executing. This is unlikely to cause a problem in practice, but it'd be a very confusing error if it does, so we can prevent it by updating the job config for the base multijob along with the rest, and having it update all the steps at once before it runs.
#18122

ihmccreery commented on hack/jenkins/job-builder-image/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+    libyaml-dev \
+    python-yaml
+RUN pip install PyYAML python-jenkins
+RUN wget https://pypi.python.org/packages/source/o/ordereddict/ordereddict-1.1.tar.gz \
```
Worth adding a comment here for what `ordereddict` is.

ihmccreery commented on hack/jenkins/job-builder-image/start.sh:
```diff
@@ -0,0 +1,8 @@
+
+# Builds and runs the kubekins-job-builder container.
+
+#docker build -t gcr.io/google_containers/kubekins-job-builder .
```
Get rid of these commented-out lines?

ihmccreery replied:
We want to use `grc.io`, and not have to `docker build` every time.

spxtr replied:
Done.

ihmccreery commented on hack/jenkins/job-builder-image/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+# This docker image runs Jenkins Job Builder for automatic job reconciliation.
+
+FROM ubuntu:14.04
```
Consider using [Busybox](https://hub.docker.com/_/busybox/)?

ihmccreery replied:
(Don't spend much time on this, it's not important for this to be optimized.  Just might be easier to maintain.  Or not.)

spxtr replied:
Ubuntu has most of the packages we need. I could look into busybox, just haven't used it before.

ihmccreery commented on hack/jenkins/update-jobs.sh:
```diff
@@ -0,0 +1,39 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Pass in the directory relative to /hack/jenkins, eg test-configs, in $1.
```
`/hack/jenkins` seems wrong.  Perhaps you meant `./hack/jenkins`?

I'd recommend making it relative to `.`; knowing that you'll descend into `./hack/jenkins` first is an extra detail I don't want to consider.

ihmccreery replied:
This comment also needs a bit more meat.  Consider a full 'usage' statement.

spxtr replied:
It's now relative to ., can add a full usage statement if you think it's worth it, but there's only one parameter.

ihmccreery commented on hack/jenkins/update-jobs.sh:
```diff
@@ -0,0 +1,39 @@
+
+pushd hack/jenkins
+
+docker exec job-builder mkdir -p /${jobs_dir}
```
if `jobs_dir` doesn't get set, then we're in `/`.  That sounds scary to me?

spxtr replied:
We have nounset set so if nothing is provided in $1 it'll crash, but even so, probably safer to put it in /jobs or something.

spxtr replied:
It's now in /kubernetes/...

ihmccreery commented on hack/jenkins/update-jobs.sh:
```diff
@@ -0,0 +1,39 @@
+set -o nounset
+set -o pipefail
+
+jobs_dir=$1
```
Quotes & brackets: `"${1}"` ([Shell Style Guide](https://google.github.io/styleguide/shell.xml))

spxtr replied:
"Don't brace-quote single character shell specials / positional parameters, unless strictly necessary or avoiding deep confusion." ;)

ihmccreery commented on hack/jenkins/update-jobs.sh:
```diff
@@ -0,0 +1,39 @@
+
+jobs_dir=$1
+
+pushd hack/jenkins
```
Quote: `"hack/jenkins"` ([Shell Style Guide](https://google.github.io/styleguide/shell.xml))

ihmccreery commented:
As we discussed, please make this as self-contained as possible: one script, no dependency on a container already running, etc.  You might reference `build/common.sh`.

spxtr commented:
This is ready for review, @ihmccreery.

ihmccreery commented on hack/jenkins/job-builder-image/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+    libyaml-dev \
+    python-yaml
+RUN pip install PyYAML python-jenkins
+# Required since jenkins-job-buider supports earlier versions of python than 2.7
```
"supports earlier versions of python than 2.7", or "doesn't support 2.7"?

spxtr replied:
Supports 2.7 and earlier.

ixdy replied:
does ubuntu14.04 use python 2.6? maybe just say that explicitly in this comment to indicate that's what you mean

spxtr replied:
No, Ubuntu 14.04 uses python 2.7, but jenkins job builder still needs this for backwards compatibility. I'll clarify the comment.

ihmccreery commented:
LGTM!

ihmccreery commented:
Undo LGTM, sorry.  Please add documentation for how to configure these things, etc.

ihmccreery commented:
Tried to run this, got:

```console
Traceback (most recent call last):
  File "/usr/local/bin/jenkins-jobs", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/dist-packages/jenkins_jobs/cmd.py", line 171, in main
    config = setup_config_settings(options)
  File "/usr/local/lib/python2.7/dist-packages/jenkins_jobs/cmd.py", line 199, in setup_config_settings
    "\n{0} is not a valid .ini file".format(conf))
jenkins_jobs.errors.JenkinsJobsException: A valid configuration file is required when not run as a test
/etc/jenkins_jobs/jenkins_jobs.ini is not a valid .ini file
```

spxtr commented:
Right, I mentioned that the credentials file needs to be in the workspace of the Jenkins job that runs it. It's just a standard [jenkins_jobs.ini](http://docs.openstack.org/infra/jenkins-job-builder/execution.html), just needs the url, the username, and the API token.

I was planning on adding this internally with the rest of the documentation that I'll be modifying/adding in the next few days, but I can add it here if that makes sense.

spxtr commented:
@ihmccreery

I updated `hack/jenkins/README.md`. It's still not enough to run this on your own, since you'll need credentials info, which won't be documented here.

ihmccreery commented on hack/jenkins/README.md:
```diff
@@ -32,5 +39,14 @@ outside this repository, it's tricky to keep documentation for it up
+
+New jobs should be specified as YAML files to be processed by [Jenkins Job
+Builder](http://docs.openstack.org/infra/jenkins-job-builder/). The YAML files
+live in `job-configs` and its subfolders. Jenkins runs Jenkins Job Builder
```
Full path please?

ixdy commented on hack/jenkins/job-builder-image/Makefile:
```diff
@@ -0,0 +1,3 @@
+all:
+	docker build -t gcr.io/google_containers/kubekins-job-builder .
```
maybe we want to have version tags like most of our other images?

ixdy commented on hack/jenkins/update-jobs.sh:
```diff
@@ -0,0 +1,39 @@
+  docker cp jenkins_jobs.ini job-builder:/etc/jenkins_jobs
+fi
+
+docker exec job-builder git pull
```
maybe also `git checkout master`? just to be safe

spxtr replied:
ok

ixdy commented on hack/jenkins/update-jobs.sh:
```diff
@@ -0,0 +1,39 @@
+    gcr.io/google_containers/kubekins-job-builder
+  # jenkins_jobs.ini contains administrative credentials for Jenkins.
+  # Store it in the workspace of the Jenkins job that calls this script.
+  docker cp jenkins_jobs.ini job-builder:/etc/jenkins_jobs
```
maybe explicitly check if this exists and fail fast if it doesn't?

spxtr replied:
sure

ixdy commented on hack/jenkins/README.md:
```diff
@@ -4,16 +4,23 @@
+* On trigger, and every half hour (which effectively means all the time, unless
+  we're failing cluster creation), e2e variants run, on the latest build assets
+  in GCS:
+  * `kubernetes-e2e-gce`: Standard GCE e2e.
```
this list is obviously very incomplete. we should at least note that?

spxtr replied:
might as well, while I'm updating it.

spxtr committed:
Allow source-controlled Jenkins job configs.

Job configs are stored in YAML files in hack/jenkins/job-configs in the
format used by Jenkins Job Builder. A docker container running on
Jenkins fetches job configs every fifteen minutes and updates Jenkins.

spxtr commented:
Thanks for the comments, @ixdy.

ixdy commented on hack/jenkins/job-builder-image/Makefile:
```diff
@@ -0,0 +1,5 @@
+TAG = 1
```
minor nit: I think most of our tags have either followed some sort of 0.N or vN convention.

ixdy commented on hack/jenkins/update-jobs.sh:
```diff
@@ -0,0 +1,45 @@
+if ! docker inspect job-builder &> /dev/null; then
+  # jenkins_jobs.ini contains administrative credentials for Jenkins.
+  # Store it in the workspace of the Jenkins job that calls this script.
+  if [[ -e jenkins_jobs.ini ]]; then
```
probably would have inverted the logic so as to avoid nesting many levels deep
